### ISSUE TITLE
Procedural cubemaps?

### ISSUE THREAD

**Issue by evertvorster** [vegastrike/Vega-Strike-Engine-Source#658]
Hey there. I know it's been a long time, but this is something that still bothers me. 
On higher resolution monitors, the backgrounds are quite pixellated. 
Also, the cube maps form a significant part of the download of the assets. 

Would it not make sense to procedurally produce the cubemaps?
This website produces them in a browser, so it can't be that CPU intensive:
https://tools.wwwtyro.net/space-3d/index.htm

From what it looks like, if you give the program the same seeds, it produces the same cube map, so there would be no need to save them, only the seed. 
This is something that can be saved in the save file too...

**Comment by stephengtuggy** [comment]
Yes, it's a great idea. Good to know also that only the seed would need to be saved in the save file.

**Comment by BenjamenMeyer** [comment]
interesting... @Loki1950 thoughts since you're more aware of the assets stuff than most, especially regarding the VS history

**Comment by evertvorster** [comment]
There is definitely a trade-off. We will lose the ray-traced starfields that contributed a lot to the eye candy of the early VS. Unfortunately, when you ray trace them at a high enough resolution to look good on high resolution monitors, they are gigantic. That is, if you manage to get the ray tracer working. I only managed to get a few of the star fields ray-traced at higher resolutions. 

There are at least three examples of procedural rendering of star fields and nebulae with a specific seed online. As another example, Oolite also procedurally generates the star fields. Unfortunately I know nothing about writing OpenGL code, so I don't even know how big of a task it is to implement a star field renderer.

**Comment by Loki1950** [comment]
We would need to experiment to see if there is an optimal setting for our purpose ie: no pixelation ,fast generation,light memory footprint. I do like the concept of procedural backgrounds but as evert has said the final maps are huge so just saving the seed would be great but will that be ok to generate on the fly or pregenerate and save it comes down to being a performance issue.

**Comment by evertvorster** [comment]
The generation part in the browser is pretty quick, so I am assuming that the starfield  can be generated every time it is needed. Or, maybe a little before... I know that quite a few of the similar type games play some sort of video that shows the time in hyperspace between star systems while the new star field is being generated. 
Then the generated starfield does not need to hit the disk at all, and only the seed is saved in the savegame. 

Conceivably, the first time the player visits a system, it gets a random seed, which is then saved for further re-use. 

What is interesting to note as well, at least on the web-based generation,  is that you can pick nebulae on or off, so if a groups of systems are supposed to be in a nebula, all the systems in that area would then have nebula on. Also, and this is just a hunch, there may be a seed for the starfield, and a seed for the nebula, so you can have the same nebula and a different star fields, if you move between systems that are close together.

There may also be an option to pick color of the nebula separately from the shape of the nebula. In this manner, a group of systems in the same nebula can have a consistent color of the nebula, but the shape changes for each system. 

So, three seeds may need to be saved. A seed for the star field, a seed for the color of the nebula, and a seed for the shape of the nebula. 
With this there will be a nice cohesive visuals to the universe, and a very nice uptick in eye candy for VS.

All this would depend on the implementation. 
Which brings us to the question, who has the skill and inclination to whip up a star field generator for VS in OpenGL?

Last point to mention, this will probably be a non-backwards compatible change to the save files. You can probably use a save file with no seed saved, and generate them on the fly as you travel back to places you have been before, but that seed saved in the save file would probably confuse earlier versions of the engine. 

Also, this may break compatibility with Privateer, unless we have the code check for existing cube maps, and load them, otherwise start generating the star maps procedurally.

**Comment by evertvorster** [comment]
Oh, just in case I have not shared it anywhere else, here is a link to a set of high-resolution cube maps that works with the current engine. This is to show what Vega Strike looks like with high resolution cube maps.

https://drive.google.com/drive/folders/1AmNTCTmfcTeVDEYRMo6KZfTgLol3RzgZ?usp=sharing

Most were generated with the online utility that I referenced earlier. This set also includes the high resolution ray traces of the maps that I managed to get to ray-trace. 
Lastly, Sol has the actual star map as it's seen from Earth today.

In order to install it, just extract the files (They expand to about 1.9Gb) and overwrite the files of the same name in game assets.

**Comment by BenjamenMeyer** [comment]
if we can generate, then perhaps we save the seed with the game data and do the generation of systems for the story line on install; non-story line systems could be a mix of stored seed and random seed based on what we provide. This would go a long way towards having a common universe as a base; it'd also be extremely helpful when we get back to networked game play where we could share the seeds.

**Comment by evertvorster** [comment]
Thinking about this today. 
Games like Oolite generate their starfields dynamically, but it does not feel like you are flying around in a galaxy, ie: most starfields look the same. 

I do believe that the skyboxes in VegaStrike tried to simulate visiting various parts of our Galaxy by varying the nebula size and color, and also the number and density of the stars present. This is why which starfield is loaded when you visit a location in Vega Strike is encoded into the save file. 

A long term goal might be to generate this galaxy once, and then procedurally generate the skyboxes depending on where in the galaxy you find yourself.  This should give you a heightened sense of realism. 

Variables like star density, and nebula size and color can then be pulled from this generated galaxy. 

In any case, this is a pie in the sky thing. 
For now, I still have a copy of the higher resolution backgrounds that I am quite willing to share with the Vega Strike community. Unfortunately these files are rather large at around 2Gb, and putting them into the Master Assets repo would make the size of the repo balloon needlessly. 

I've been looking at https://docs.github.com/en/repositories/working-with-files/managing-large-files
This details a way of storing large binary files, but unfortunately this is not a free service for the size of files that I have already created, and also not a long term solution as we want to be moving to procedurally generating the starfields anyways. 

So, the best I can come up with right now for supporting high resolution backgrounds in Vega Strike is to keep sharing those high resolution backgrounds on my google drive and providing instructions on how to install them. 

The install instructions are quite simple:

1. Download the file from this link:
https://drive.google.com/file/d/14RuGdMZFQ_mZGykrUtBSAR4aZMPigkqs/view?usp=sharing
It's only 250mb compressed tarball.

2. Extract the file. 
You will get a directory called 4096_Textures, and in there you will have all the .cube files that have the higher resolution textures. They have exactly the same names as the textures that is in /usr/share/vegastrike/textures/backgrounds/

3. Simply copy over the textures to where you have VegaStrike assets installed, overwriting the textures that are there. 
As an added benefit, the starfield in Sol is really the starfield you would see if you looked up on a starry night on Earth. 

What are the thoughts of sharing this instruction on the main readme page of VegaStrike github?

**Comment by evertvorster** [comment]
I've been asking ChatGPT about a lot of things, and while it is incredibly knowledgeable, it does sometimes need a little guidance on details. Not knowing anything about any of the technologies mentioned, this would require me to research each step and of course do the coding and effort myself to avoid any copyright issues. 

Here is what ChatGPT thinks about this subject: 

# Upgrading the Vega Strike Skybox System

Vega Strike’s pre‑rendered cubemap skyboxes were a clever solution in the early‑2000s era of fixed‑function OpenGL, but on today’s hardware there are several ways you can make them lighter, crisper, and more flexible—without throwing away the existing tool‑chain or its famous moddability.

## Project status (as of 2025‑05‑26)

* **Active modernisation work:** The main engine repo is alive on GitHub; the latest tagged build is `v0.9.1` (May 26 2025) and the roadmap shows continued refactors toward cleaner C++17 code and better Windows/macOS builds.
* **Python bindings:** being migrated from Boost.Python to **pybind11**, which simplifies build deps.
* **Renderer:** still largely OpenGL 2.x; long‑term plan is to move to a core profile 3.3+/4.x path.

---

## Upgrade options for the skybox system

| Approach | What changes | Pros | Cons |
|----------|--------------|------|------|
| **A. Re‑bake cubemaps at higher res / HDR** | Keep current lookup logic, move assets to 8–16 k HDR **KTX2** with Basis U compression | Quick win; zero C++ changes; re‑uses modding pipeline | Larger textures; still static |
| **B. Procedural starfield shader + lightweight nebula LUTs** | Replace six‑face cube with a fullscreen draw call that samples a star‑catalog SSBO and blends a low‑res nebula cube or 3‑D noise | Tiny GPU memory footprint; infinite resolution; parallax‑correct | Requires GLSL 3.30+; need star‑catalog preprocessing |
| **C. Hybrid: on‑demand dynamic cubemap** | Render skybox once per sector in an off‑screen FBO, cache to disk | High detail and consistency with minimal startup cost | Additional code: render‑to‑cubemap, cache invalidation |

---

## Recommended migration path

### Short term (weeks)

1. Ship **Approach A**: convert existing PNG/DDS faces to KTX2 (BasisU super‑compression).  
2. Patch the texture loader to prefer `.ktx2`, fall back to legacy formats (~50 LoC).

### Medium term (months)

* Add an **optional procedural starfield** (Approach B) behind a `--experimental-render skybox=procedural` CLI/config flag.  
* Provide `tools/generate_star_catalog.py` to down‑sample **Gaia EDR3** stars (e.g., mag ≤ 8.0) and pack into a binary blob.

### Long term (post‑1.0)

* Once the renderer hits GL 4.x / Vulkan, explore volumetric nebulae via compute‑shader ray‑marching or cached cubemaps (Approach C) for dense regions like planetary rings.

---

## Tooling & pipeline tips

* **Blender 4.x** + the *Cube Map Baker* add‑on exports perfect 6‑face cubes already aligned to Vega Strike’s coordinates.  
* Bulk‑convert to KTX2:

  ```bash
  for dir in textures/skybox/*; do
      toktx --bcmp --2d -q etc1s "$(basename "$dir").ktx2" "$dir"/*.png
  done
  ```

* For procedural stars: use an open‑source **gaia‑hammer** (or similar) script to filter and pack Gaia EDR3 into a binary SSBO (magnitude, colour index, proper motion) so stars can twinkle during time‑compression.

---

## Integration checklist

- **Loader upgrade:** add KTX2 branch (`gfx_texture.cpp`) using `glCompressedTexImage2D`.  
- **Config:**  

  ```ini
  [Graphics]
  Skybox = cubemap      ; cubemap | procedural
  SkyboxCompression = KTX2
  ```

- **Mod packager:** update to detect `.ktx2` and auto‑fill metadata.  
- **Documentation:** write wiki page “Creating HDR Skyboxes for v0.9+” with example Blender scene and conversion scripts.

---

## Next steps

* Need a **step‑by‑step guide** for converting a single skybox?  
* Or a **prototype GLSL starfield shader** to drop into `src/gfx/`?  
* Let me know and I can supply code snippets or pipeline walkthroughs.

**Comment by stephengtuggy** [comment]
`Active modernisation work` -- That part is accurate.

Moving Python bindings to pybind11 -- Really? First I've heard of it LOL

And what the heck is a `gaia-hammer` script, and how does it help stars `twinkle during time-compression`? LOL

Overall, ChatGPT's response here seems to be about half accurate and half glib nonsense.

**Comment by evertvorster** [comment]
Yeah, it dreams up quite a few things. This may just be because what I normally ask it to do is so well outside of its training that it comes up with some pretty weird stuff. 

Saying all that, I use it more like a search engine than an actual advisor, and it does sometimes find some pretty creative things on the internet. 

In the end, though, that stuff about using a more modern cubemap compression sounds interesting, and it is probably right about using more modern OpenGL.

**Comment by stephengtuggy** [comment]
> Yeah, it dreams up quite a few things. This may just be because what I normally ask it to do is so well outside of its training that it comes up with some pretty weird stuff.
> 
> Saying all that, I use it more like a search engine than an actual advisor, and it does sometimes find some pretty creative things on the internet.
> 
> In the end, though, that stuff about using a more modern cubemap compression sounds interesting, and it is probably right about using more modern OpenGL.

Agreed.

**Comment by BenjamenMeyer** [comment]
> A long term goal might be to generate this galaxy once, and then procedurally generate the skyboxes depending on where in the galaxy you find yourself. This should give you a heightened sense of realism.

Yes, we need to be able to do that in a way that can support multi-player too, even for our whole universe generation. This is probably just one aspect of that.
